spirv_msl.cpp

```diff
@@ -138,6 +165,92 @@ void CompilerMSL::localize_global_variables()
 	}
 }
 
+// For any global variable accessed directly by a function,
+// extract that variable and add it as an argument to that function.
+void CompilerMSL::extract_global_variables_from_functions()
+{
+
+    // Uniforms
+    std::set<uint32_t> global_var_ids;
+    for (auto &id : ids)
+    {
+        if (id.get_type() == TypeVariable)
+        {
+            auto &var = id.get<SPIRVariable>();
+            if (var.storage == StorageClassInput ||
+                var.storage == StorageClassUniform ||
+                var.storage == StorageClassUniformConstant ||
+                var.storage == StorageClassPushConstant)
+                    global_var_ids.insert(var.self);
+        }
+    }
+
+    std::set<uint32_t> added_arg_ids;
+    std::set<uint32_t> processed_func_ids;
+    extract_global_variables_from_function(execution.entry_point, added_arg_ids, global_var_ids, processed_func_ids);
+}
+
+// MSL does not support the use of global variables for shader input content.
+// For any global variable accessed directly by the specified function, extract that variable,
+// add it as an argument to that function, and the arg to the added_arg_ids collection.
+void CompilerMSL::extract_global_variables_from_function(uint32_t func_id,
+                                                         std::set<uint32_t>& added_arg_ids,
+                                                         std::set<uint32_t>& global_var_ids,
+                                                         std::set<uint32_t>& processed_func_ids)
+{
+    // Avoid processing a function more than once
+    if ( processed_func_ids.find(func_id) != processed_func_ids.end() )
+        return;
+
+    processed_func_ids.insert(func_id);
+
+    auto& func = get<SPIRFunction>(func_id);
+
+    // Recursively establish global args added to functions on which we depend.
+    for (auto block : func.blocks)
+    {
+        auto &b = get<SPIRBlock>(block);
+        for (auto &i : b.ops)
+        {
+            auto ops = stream(i);
+            auto op = static_cast<Op>(i.op);
+
+            switch (op) {
+                case OpLoad:
+                case OpAccessChain: {
```

Comment:
Uncertain. Haven't encountered need yet. Will leave for now.  OpStore might require different arg binding.